examples/redbook/03_feedback/src/main_loop.adb

```diff
@@ -0,0 +1,366 @@
+      GL.Objects.Vertex_Arrays.Draw_Arrays
+        (Mode  => GL.Types.Points, First => 0,
+         Count => GL.Types.Size (UInt'Min (Num_Points, Frame_Count)));
+      GL.Objects.Programs.End_Transform_Feedback;
+
+      GL.Objects.Vertex_Arrays.Bind (GL.Objects.Vertex_Arrays.Null_Array_Object);
+
+      if Frame_Count > 5000 then
+         Frame_Count := 0;
+      else
+         Frame_Count := Frame_Count + 1;
+      end if;
+
+   exception
+      when  others =>
+         Put_Line ("An exception occurred in Main_Loop.Display.");
+         raise;
+   end Display;
+
+   --  ------------------------------------------------------------------------
+   --  Load_PV_Buffer loads Transform_Feedback_Buffer with data points
+   procedure Load_PV_Buffer is
+      use GL.Objects.Buffers;
+      use GL.Types.Singles;
+      Buffer_Pointer : PV_Buffer_Package.Pointer;  --  to PV_Buffer_Array
+      Velocity       : Vector3;
+   begin
+      Map_PV_Buffer (Transform_Feedback_Buffer, GL.Objects.Write_Only, Buffer_Pointer);
+      for B_Index in 1 .. Num_Points loop   --  j
+         Velocity := Random_Vector;
+         Buffer_Pointer.Position := To_Vector4 (Velocity) + (-0.5, 40.0, 0.0, 0.0);
+         Buffer_Pointer.Position := To_Vector4 (0.1 * Velocity) + (-0.5, 0.2, 0.0, 0.0);
+         Buffer_Pointer.Velocity := (Velocity (GL.X), 0.3 * Velocity (GL.Y),
+                         0.3 * Velocity (GL.Z));
+         PV_Buffer_Package.Increment (Buffer_Pointer);
+      end loop;
+      Unmap (Transform_Feedback_Buffer);
+   exception
+      when  others =>
+         Put_Line ("An exception occurred in Main_Loop.Load_PV_Buffer.");
+         raise;
+   end Load_PV_Buffer;
+
+   --  ------------------------------------------------------------------------
+
+   function Setup return Boolean is
+      use GL.Objects.Buffers;
+      use GL.Objects.Programs;
+      use GL.Objects.Shaders;
+      use Program_Loader;
+      VBM_Result     : Boolean := False;
+      Varyings       : constant String := "position_out,velocity_out";
+      Varyings_2     : constant String := "world_space_position";
+      --  Commented out to allow debug of Transform_Feedback_Varyings problem
+--        Name           : String (1 .. 30);
+--        Length         : GL.Types.Size := 99;
+      V_Length       : GL.Types.Size := 99;
+      Max_Length     : Int;
+      V_Type         : Buffer_Mode;
+   begin
+      VAO (1).Initialize_Id;
+      VAO (1).Bind;
+      VAO (2).Initialize_Id;
+      VAO (2).Bind;
+      Render_VAO.Initialize_Id;
+      GL.Objects.Vertex_Arrays.Bind (Render_VAO);
+
+      for index in VBO'Range loop
+         VBO (index).Initialize_Id;
+         Transform_Feedback_Buffer.Bind (VBO (index));
+      end loop;
+
+      Update_Program := Program_From
+        ((Src ("src/shaders/update_vertex_shader.glsl", Vertex_Shader),
+         Src ("src/shaders/update_fragment_shader.glsl", Fragment_Shader)));
+
+      GL.Objects.Programs.Use_Program  (Update_Program);
```

Comment:
Oh!
My mistake I had thought that `Program_From` included linking!

Reply:
GDB seems to work on my system.
I just did a quick check so maybe not all functionality works.
Any suggestions as to what I might try?

Reply:
Well, I would be interested in how the program behaves on your machine if you remove that line. It might be that `Use_Program` will lock the program somehow so linking `Transform_Feedback_Varyings` into it afterwards will not work.

Reply:
Deleting the  ` GL.Objects.Programs.Use_Program  (Update_Program);` that follows the `Program_From` instruction seems to make no difference as it still produces:
~~~
V_Type: INTERLEAVED_ATTRIBS   V_Length:  2   Max Length:  0
Setup, Update_Program Link failed
ERROR: Could not find transform feedback binding for 'position_out.'
ERROR: Could not find transform feedback binding for 'velocity_out.'

An exception occurred in Main_Loop.Setup.
An exception occurred in Main_Loop.
An exception occurred in Transform_Feedback.
raised GL.ERRORS.INVALID_OPERATION_ERROR : gl-raise_exception_on_opengl_error.adb:10
~~~
I was a bit surprised by this as I thought that I had tried it previously with a different result.
I had thought of trying it with the `link` step removed from `Program_From` and not attempting linking until after `Transform_Feedback_Varyings`.
I'll try that later on today and let you know the result.